### DISCUSSION TITLE
Welcome to indipydriver Discussions!

### DISCUSSION THREAD

**Discussion by bernie-skipole** [bernie-skipole/indipydriver#3] [Announcements]
## 👋 Welcome!

Anything to say about indipydriver, put it here.

**Comment by RhoLuke** [comment]
Awsome project really!
It's the Alpyca brother for INDI, amazing.
I was looking exactly for something like this because is getting more frequent now to know better python than cpp, so this kind of bindings are just gold!

Btw I was wondering, it is possible to use the `pyindiserver` together with the nominal `indiserver`. Let's make an example of having two RPis, and we want to create a chain like the following 

pyindiserver->indiserver

Is it possible to chain the pyindiserver like a standard indiserver? (I.e. `indiserver [local drivers] "remote driver"@secondo_rpi`

**Reply by bernie-skipole** [reply]
Hi there. Yes it is possible, the IPyServer class has an 'add_remote' method which makes a call to a remote INDI service, and broadcasts instrument traffic to it. For example, if the remote server is running on 'remotepi' then:

server = IPyServer(*localdrivers)
server.add_remote(host="remotepi", port=7624)
asyncio.run(server.asyncrun())

It also has an add_exdriver method for connecting executable drivers, rather than IPyDriver instances, so should be able to work with existing INDI drivers.

At least I hope it does! Much depends on my interpretation of the INDI spec matching other peoples interpretation. 

For further info see

[readthedocs ipyserver](https://indipydriver.readthedocs.io/en/latest/usage/ipyserver.html)

**Reply by bernie-skipole** [reply]
Correction to my last note, I've just read on the indilib web site:

"By default, the INDI server only accepts connections from the local machine. For secure remote access, consider using SSH tunneling."

I could not find how this default is overridden, so if indiserver is remote, then an SSH tunnel would need setting up, forwarding the port to localhost on the remote machine.

The IPyServer class also defaults to localhost - its signature is:

IPyServer(*drivers, host='localhost', port=7624, maxconnections=5)

Setting the host argument to '0.0.0.0' will make it listen on all ip interfaces. However using localhost with SSH tunneling is obviously a good idea, and should certainly be used across any insecure network.
